pkg/controllers/management/nodepool/nodepool.go

```diff
@@ -56,77 +56,95 @@ func Register(ctx context.Context, management *config.ManagementContext) {
 
 	// Add handlers
 	p.NodePools.AddLifecycle(ctx, "nodepool-provisioner", p)
-	management.Management.Nodes("").AddHandler(ctx, "nodepool-provisioner", p.machineChanged)
+	management.Management.Nodes("").AddHandler(ctx, "nodepool-provisioner", p.nodeChanged)
 }
 
+// Create does a noop for nodepool.
 func (c *Controller) Create(nodePool *v3.NodePool) (runtime.Object, error) {
 	return nodePool, nil
 }
 
+// needsReconcile checks the nodepool to see if reconciliation is necessary.
 func (c *Controller) needsReconcile(nodePool *v3.NodePool, nodes []*v3.Node) bool {
 	changed, _, err := c.createOrCheckNodes(nodePool, nodes, true)
 	if err != nil {
-		logrus.Debugf("[nodepoool] error checking pool for reconciliation: %s", err)
+		logrus.Debugf("[nodepool] error checking pool for reconciliation: %s", err)
 	}
 
 	return changed
 }
 
-func (c *Controller) reconcile(nodePool *v3.NodePool, nodes []*v3.Node) {
+// reconcile reconciles the nodepool provided if necessary
+func (c *Controller) reconcile(nodePool *v3.NodePool, nodes []*v3.Node) (runtime.Object, error) {
 	_, qty, err := c.createOrCheckNodes(nodePool, nodes, false)
 	if err != nil {
-		logrus.Errorf("[nodepool] reconcile error, create or check nodes: %s", err)
+		return nodePool, fmt.Errorf("[nodepool] reconcile error, create or check nodes: %s", err)
 	}
 
+	// In certain cases, the nodepool quantity can actually decrease through a `createOrCheckNodes` call, like
+	// for scaling down a specific node. In this case, we need to update the nodePool object with the latest
+	// quantity so that on a future reconciliation loop, we do not recreate a node that was specifically scaled down.
 	if qty != nodePool.Spec.Quantity {
 		nodePool.Spec.Quantity = qty
 	}
 
-	_, err = c.setReconcileAnnotation(nodePool, "")
+	nodePool, err = c.updateNodePoolAnnotationAndSpec(nodePool, "", true)
 	if err != nil {
-		logrus.Errorf("[nodepool] error updating reconcile annotation to updated: %s", err)
+		return nodePool, fmt.Errorf("[nodepool] error updating reconcile annotation to updated: %s", err)
 	}
+	return nodePool, nil
 }
 
+// Updated reconciles a nodepool if necessary
 func (c *Controller) Updated(nodePool *v3.NodePool) (runtime.Object, error) {
+	if nodePool != nil {
+		logrus.Tracef("[nodepool] Updated called for nodepool %s: %+v", nodePool.Name, nodePool)
+	}
 	obj, err := v32.NodePoolConditionUpdated.Do(nodePool, func() (runtime.Object, error) {
 		anno, _ := nodePool.Annotations[ReconcileAnnotation]
+		logrus.Debugf("[nodepool] nodepool %s reconcile annotation value was: %s", nodePool.Name, anno)
 		if anno == "" {
-			nodes, err := c.NodeLister.List(nodePool.Namespace, labels.Everything())
+			nodes, err := c.nodes(nodePool)
 			if err != nil {
 				return nodePool, err
 			}
-			go c.deleteBadNodes(nodes)
```

Comment:
deleteBadNodes took care of deleting nodes if they failed to came up in 2 mins. once we return from our nodePool logic now, we won't get other nodePool syncs to delete these nodes. is this something we need to account for?

Reply:
This might break our expected usecase of 
`
After 2-3 minutes the stopped nodes should be removed and replaced from Rancher, additional nodes beyond the stopped count should not be removed or created
`

Reply:
deleteBadNodes the way I interpreted it, set a 2 minute countdown for deletion if the conditions didn’t exist at time of loop reconcile. It did not recheck conditions right before deletion. I figure the node will eventually get deleted, but not at the guaranteed 2 minute mark. I could be wrong on my original interpretation though

Reply:
ah nvm, I see the check for conditions is now moved to create/check nodes func. That should cover it